proof/courier.go

```diff
@@ -854,6 +926,199 @@ func (h *HashMailCourier) SetSubscribers(
+		})
+		if err != nil {
+			return fmt.Errorf("error inserting proof into "+
+				"universe courier service: %w", err)
+		}
+	}
+
+	return err
+}
+
+// ReceiveProof attempts to obtain a proof file from the courier service. The
+// final proof in the target proof file is identified by the given locator.
+func (c *UniverseRpcCourier) ReceiveProof(ctx context.Context,
+	originLocator Locator) (*AnnotatedProof, error) {
+
+	// In order to reconstruct the proof file we must collect all the
+	// transition proofs that make up the main chain of proofs. That is
+	// accomplished by iterating backwards through the main chain of proofs
+	// until we reach the genesis point (minting proof).
+
+	// We will update the locator at each iteration.
+	loc := originLocator
+
+	// revProofs is a slice of transition proofs ordered from latest to
+	// earliest.
+	var revProofs []Proof
+
+	for {
+		universeID := unirpc.MarshalUniverseID(*loc.AssetID)
+		assetKey := unirpc.MarshalAssetKey(
+			*loc.OutPoint, &loc.ScriptKey,
+		)
+		universeKey := unirpc.UniverseKey{
+			Id:      &universeID,
+			LeafKey: &assetKey,
+		}
+
+		resp, err := c.client.QueryProof(ctx, &universeKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decode transition proof from query response.
+		proofBlob := resp.AssetLeaf.IssuanceProof
+		var transitionProof Proof
+		if err := transitionProof.Decode(
+			bytes.NewReader(proofBlob),
+		); err != nil {
+			return nil, err
+		}
+
+		revProofs = append(revProofs, transitionProof)
+
+		// Break if we've reached the genesis point (the asset is the
+		// genesis asset).
+		if transitionProof.Asset.HasGenesisWitness() {
+			break
+		}
+
+		// Update locator with principal input to the current outpoint.
+		prevWitnesses := transitionProof.Asset.PrevWitnesses
+		if len(prevWitnesses) == 0 {
+			return nil, fmt.Errorf("no previous witnesses found")
+		}
+		prevID := prevWitnesses[0].PrevID
+
+		// Parse script key public key.
+		scriptKeyPubKey, err := btcec.ParsePubKey(prevID.ScriptKey[:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse script key "+
+				"public key from Proof.PrevID: %w", err)
+		}
+		loc.ScriptKey = *scriptKeyPubKey
+
+		loc.AssetID = &prevID.ID
+		loc.OutPoint = &prevID.OutPoint
+	}
+
+	// Append proofs to proof file in reverse order to their collected
+	// order.
+	proofFile := &File{}
+	for i := len(revProofs) - 1; i >= 0; i-- {
+		err := proofFile.AppendProof(revProofs[i])
+		if err != nil {
+			return nil, fmt.Errorf("error appending proof to "+
+				"proof file: %w", err)
+		}
+	}
+
+	// Encode the full proof file.
+	var buf bytes.Buffer
+	if err := proofFile.Encode(&buf); err != nil {
+		return nil, fmt.Errorf("error encoding proof file: %w", err)
+	}
+	proofFileBlob := buf.Bytes()
+
+	return &AnnotatedProof{
+		Locator: originLocator,
+		Blob:    proofFileBlob,
+	}, nil
+}
+
+// SetSubscribers sets the subscribers for the courier. This method is
+// thread-safe.
+func (c *UniverseRpcCourier) SetSubscribers(
```

Comment:
We're currently not using anything subscriber related. Can we just remove this, the map and the mutex? Or is it part of an interface?

Reply:
It's part of the proof courier interface. I think the hashmail courier's backoff procedure uses this to publish events.